libs/image/src/Image.cpp

```diff
@@ -62,15 +60,40 @@ void Image::subset(Image const& image,
     mOwnedData.release();
     mWidth = w;
     mHeight = h;
-    mFlags ^= flags & FLIP_XY;
     mBpr = image.mBpr;
     mBpp = image.mBpp;
     mChannels = image.mChannels;
     mData = static_cast<uint8_t*>(image.getPixelRef(x, y));
 }
 
-void Image::setFlags(uint32_t flags) {
-    mFlags = flags;
+void Image::flip(uint32_t flags) {
+    // We verified that these lambdas get inlined by clang in release builds.
+    auto getptr = [this](size_t x, size_t y) -> uint8_t* {
+        return static_cast<uint8_t*>(mData) + y*mBpr + x*mBpp;
+    };
+    auto getref = [this, getptr](size_t x, size_t y) -> uint8_t& {
+        return *(getptr(x, y));
+    };
+    if (flags & Image::FLIP_Y) {
+        std::unique_ptr<uint8_t[]> tmp(new uint8_t[mBpr]);
+        uint8_t* ptmp = tmp.get();
+        for (size_t row = 0; row < mHeight / 2; ++row) {
+            uint8_t* a = getptr(0, row);
+            uint8_t* b = getptr(0, mHeight - 1 - row);
+            memcpy(ptmp, a, mBpr);
+            memcpy(a, b, mBpr);
+            memcpy(b, ptmp, mBpr);
+        }
+    }
+    // Our hflip implementation is inefficient, but it's never invoked in the renderer.
+    if (flags & Image::FLIP_X) {
+        for (size_t row = 0; row < mHeight; ++row) {
```

Comment:
use instead: for (size_t row = 0, rc = mHeight; row < rc; ++row)

Reply:
Done.